private/buf/cmd/buf/command/push/push_test.go

```diff
@@ -0,0 +1,259 @@
+}
+
+var _ registryv1alpha1connect.PushServiceHandler = (*mockPushService)(nil)
+
+func newMockPushService(t *testing.T) *mockPushService {
+	return &mockPushService{
+		t:         t,
+		callbacks: make(map[string]func(*registryv1alpha1.PushRequest)),
+		called:    make(map[string]struct{}),
+		resp:      make(map[string]*registryv1alpha1.PushResponse),
+	}
+}
+
+// Push pushes.
+func (m *mockPushService) Push(
+	_ context.Context,
+	req *connect_go.Request[registryv1alpha1.PushRequest],
+) (*connect_go.Response[registryv1alpha1.PushResponse], error) {
+	owner := req.Msg.Owner
+	cb, ok := m.callbacks[owner]
+	if ok {
+		cb(req.Msg)
+		m.called[owner] = struct{}{}
+	}
+	resp := m.resp[owner]
+	if resp != nil {
+		return &connect_go.Response[registryv1alpha1.PushResponse]{
+			Msg: resp,
+		}, nil
+	}
+	return nil, errors.New("bad request")
+}
+
+func (m *mockPushService) respond(owner string, resp *registryv1alpha1.PushResponse) {
+	m.resp[owner] = resp
+}
+
+func (m *mockPushService) callback(
+	owner string,
+	cb func(*registryv1alpha1.PushRequest),
+) {
+	m.callbacks[owner] = cb
+}
+
+func (m *mockPushService) assertAllCallbacksCalled() {
+	for k := range m.callbacks {
+		_, ok := m.called[k]
+		assert.True(m.t, ok)
+	}
+}
+
+func TestPush(t *testing.T) {
+	server, mock := pushService(t)
+	t.Cleanup(func() {
+		server.Close()
+		mock.assertAllCallbacksCalled()
+	})
+
+	testPush(
+		t,
+		"success",
+		server.URL,
+		mock,
+		&registryv1alpha1.PushResponse{
+			LocalModulePin: &registryv1alpha1.LocalModulePin{},
+		},
+		"",
+	)
+	testPush(
+		t,
+		"missing local module pin",
+		server.URL,
+		mock,
+		&registryv1alpha1.PushResponse{
+			LocalModulePin: nil,
+		},
+		"Missing local module pin",
+	)
+	testPush(
+		t,
+		"registry error",
+		server.URL,
+		mock,
+		nil,
+		"bad request",
+	)
+}
+
+func pushService(t *testing.T) (*httptest.Server, *mockPushService) {
+	mock := newMockPushService(t)
+	mux := http.NewServeMux()
+	mux.Handle(
+		registryv1alpha1connect.NewPushServiceHandler(mock),
+	)
+	return httptest.NewServer(mux), mock
+}
+
+func testPush(
+	t *testing.T,
+	desc string,
+	URL string,
+	mock *mockPushService,
+	resp *registryv1alpha1.PushResponse,
+	errorMsg string,
+) {
+	t.Helper()
+	owner := strings.ReplaceAll(desc, " ", "_")
+	mock.respond(owner, resp)
+	mock.callback(owner, func(req *registryv1alpha1.PushRequest) {
+		assert.NotNil(t, req.Module, "missing module")
+	})
+	t.Run(desc, func(t *testing.T) {
+		t.Parallel()
+		appName := "test"
+		err := appcmd.Run(
+			context.Background(),
+			app.NewContainer(
+				ammendEnv(
+					internaltesting.NewEnvFunc(t),
+					func(env map[string]string) map[string]string {
+						env["BUF_TOKEN"] = "invalid"
+						buftransport.SetDisableAPISubdomain(env)
+						injectConfig(t, appName, env)
+						return env
+					},
+				)(appName),
+				tarball(map[string][]byte{
+					"buf.yaml":  bufYAML(t, URL, owner, "repo"),
+					"foo.proto": nil,
```

Comment:
what does it mean for a file to have `nil` contents in a tarball?

Reply:
`nil` is the empty slice, hence the same as `[]byte{}`, and means an empty file. "foo.proto" and "bar.proto" are both empty files, which is fine for the tests written here. That may change if I get to testing when two proto files do not have identical content.